### ISSUE TITLE
counsel-M-x : Noticeably slower when running remotely

### ISSUE THREAD

**Issue by murphytalk** [abo-abo/swiper#1246]
When running locally (on PC) there is no slowness whatsoever, but using the same configuration to run emacs25 on a remote server is noticeably slower, tried both emacs-nox in ssh terminal and emacs X11 (using Exceed X-Server with ssh X11 tunneling).

The typical syndrome is that emacs would freeze for several and sometimes dozens of seconds or even longer when run counsel-M-x, then (sometimes) resumes. I've tried this on both our Redhat 6.5 box (Cisco UCS physical blade) and Solaris box (M-6000), same experience.

Network latency shouldn't be an issue, as the servers are in a DC only a few kilometers away and we are on the fast corp network, all the other apps just work fine.

Does this need to scan and iterate through files? One possible reason that I can think of is that on the servers my home directory is mounted from NFS, which is slower than local file system as we know ...

Very much appreciated if you can shed some light and give some advice on troubleshooting. Thanks.

PS. below is a profiling

```
- command-execute                                               22039  87%
 - call-interactively                                           22039  87%
  - funcall-interactively                                       22039  87%
   - my-M-x                                                     18096  71%
    - cond                                                      18096  71%
     - counsel-M-x                                              18096  71%
      - let*                                                    18096  71%
       - ivy-read                                               12955  51%
        - let*                                                  12955  51%
         - progn                                                12954  51%
          - progn                                               12954  51%
           - let                                                12954  51%
            - prog1                                             12933  51%
             - unwind-protect                                   12933  51%
              - let                                             12931  51%
               + unwind-protect                                 12931  51%
                - progn                                         12931  51%
                 - let*                                         12931  51%
                  - if                                          12931  51%
                   - read-from-minibuffer                        3067  12%
                    + #<lambda 0x1b6404eb6be384ec>                670   2%
                    + ivy--exhibit                                429   1%
                    + timer-event-handler                         185   0%
                    + redisplay_internal (C function)                 12   0%
                    + command-execute                               8   0%
                    + minibuffer-inactive-mode                      6   0%
                    + my-minibuffer-setup-hook                      1   0%
                    + global-font-lock-mode-check-buffers                  1   0%
                    + frame-windows-min-size                        1   0%

```

**Comment by abo-abo** [comment]
> When running locally (on PC) there is no slowness whatsoever,

Which OS and Emacs version?

> but using the same configuration to run emacs25 on a remote server is noticeably slower, tried both emacs-nox in ssh terminal and emacs X11 (using Exceed X-Server with ssh X11 tunneling).

I never run Emacs on remote, TRAMP provides everything I need. But I think it's important to have the option. And Ivy should try to be fast if the user runs Emacs on remote.

> The typical syndrome is that emacs would freeze for several and sometimes dozens of seconds or even longer when run counsel-M-x, then (sometimes) resumes. I've tried this on both our Redhat 6.5 box (Cisco UCS physical blade) and Solaris box (M-6000), same experience.

Is it only `counsel-M-x`, or are other commands affected too?

**Comment by basil-conto** [comment]
> Does this need to scan and iterate through files?

Yes, currently each `counsel-M-x` invocation hits the file system on up to two occasions: see the subroutine `counsel--M-x-externs` which tries to detect the presence of one of the external packages `amx` or `smex`. In the worst case, if neither of these packages is installed, the entire `load-path` will be searched twice for their presence.

> One possible reason that I can think of is that on the servers my home directory is mounted from NFS, which is slower than local file system as we know ...

It's not immediately clear from your profiler report that this is indeed the issue, but it's my best guess at a potentially expensive file system operation involved in vanilla `counsel-M-x`.

As someone who doesn't use `amx` or `smex`, I would also ideally like to skip these redundant `require`s :).
